supporter-product-data/src/main/scala/com/gu/lambdas/UpdateDynamoLambda.scala

```diff
@@ -103,19 +107,34 @@ object UpdateDynamoLambda extends StrictLogging {
   def writeBatch(list: List[(SupporterRatePlanItem, Int)], dynamoDBService: DynamoDBService, alarmService: AlarmService) = {
     val futures = list.map {
       case (supporterRatePlanItem, index) =>
-        logger.info(
-          s"Attempting to write item index $index - ${supporterRatePlanItem.productRatePlanName} " +
-            s"rate plan with term end date ${supporterRatePlanItem.termEndDate} to Dynamo")
-        dynamoDBService
-          .writeItem(supporterRatePlanItem)
-          .recover {
-            // let's alarm and keep going if one insert fails
-            case error: Throwable =>
-              logger.error(s"An error occurred trying to write item $supporterRatePlanItem, at index $index", error)
-              alarmService.triggerDynamoWriteAlarm
-          }
-
+        logger.info(s"Attempting to write item index $index - ${supporterRatePlanItem.productRatePlanName} to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")
+        dynamoDBService.writeItem(supporterRatePlanItem)
     }
     Future.sequence(futures)
   }
+
+  def batchItemsWhichCanUpdateConcurrently(items: List[(SupporterRatePlanItem, Int)]): List[List[(SupporterRatePlanItem, Int)]] = {
+    // 'Batch' supporterRatePlanItems up into groups which we can update in Dynamo concurrently.
+    // For this to be safe we need to make sure that no group has more than one item with the same subscription name in it because if it does
+    // then order of execution is important and we can't guarantee this with parallel executions.
+
+    def batchAlreadyHasAnItemForThisSubscription(batch: ListBuffer[(SupporterRatePlanItem, Int)], subscriptionName: String) =
+      batch.exists {
+        case (item, _) => item.subscriptionName == subscriptionName
+      }
+
+    // Using mutable state for performance reasons, it is many times faster than the immutable version
```

Comment:
Shouldn't be the case fundamentally, I don't think it's the mutability that makes a difference here.
were you appending to a linked `List` as that would involve copying the whole list so be O(n) each time and O(n^2) overall i.e. very bad
The usual way would be prepend which is O(1) and then reverse the list if necessary at the end which would be O(n) meaning the whole thing would be O(n) i.e. good
 ListBuffer append would be O(log n) which would make it O(n log n) overall which is also generally acceptable.
If you really want to append rather than prepend and reverse, then immutable Vector is also log n for append.